### DISCUSSION TITLE
ProgressStream only returning 0

### DISCUSSION THREAD

**Discussion by Sparky54** [pschatzmann/arduino-audio-tools#913] [Q&A] [answered]
I'm trying to use progressStream by passing the I2SStream to it but in my serial monitor it's straight 0%'s. Here is my code:

#include "AudioTools.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include <SPI.h>

#define PIN_AUDIO_KIT_SD_CARD_CS 22
#define PIN_AUDIO_KIT_SD_CARD_MISO 19
#define PIN_AUDIO_KIT_SD_CARD_MOSI 23
#define PIN_AUDIO_KIT_SD_CARD_CLK 18

const char *startFilePath = "/";
const char *ext = "mp3";
AudioSourceSDFAT source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);
ProgressStream progressStream(i2s);

void printMetaData(MetaDataType type, const char *str, int len) {
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);

  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_bck = 26;
  cfg.pin_ws = 25;
  cfg.pin_data = 27;
  i2s.begin(cfg);

  player.setMetadataCallback(printMetaData);
  progressStream.begin();
  player.begin();
}

void loop() {
  player.copy();

  Serial.print("Progress: ");
  Serial.print(progressStream.percentage());
  Serial.println("%");
  
  delay(1000);
}

Any ideas on where I'm messing up? Thanks in advance :)

**Comment by pschatzmann** [comment, marked as answer]
You forgot to provide the file size! You can't calculate a percentage if the total is unknown!

**Reply by Sparky54** [reply]
Sorry, I'm unclear as how to get the file size

**Comment by pschatzmann** [comment]
This has nothing to do with my framework: Google is your friend and I am sure it should not bee to difficult to find out how to determine the size of a file

**Comment by Sparky54** [comment]
Ok, so I switched from AudioSourceSDFAT to AudioSourceSD as the SD library has a method to get file size, forked your library, 
added my own method in AudioSourceSD.h: `size_t getFileSize() { return file.size(); }`

then implemented all of this in my program: 

```
#include "AudioTools.h"
#include "AudioLibs/AudioSourceSD.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include <SPI.h>

#define PIN_AUDIO_KIT_SD_CARD_CS 22
#define PIN_AUDIO_KIT_SD_CARD_MISO 19
#define PIN_AUDIO_KIT_SD_CARD_MOSI 23
#define PIN_AUDIO_KIT_SD_CARD_CLK 18

const char *startFilePath = "/";
const char *ext = "mp3";
AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);
ProgressStream progressStream(i2s);
File file;
size_t fileSize;

void printMetaData(MetaDataType type, const char *str, int len) {
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);

  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_bck = 26;
  cfg.pin_ws = 25;
  cfg.pin_data = 27;
  i2s.begin(cfg);

  player.setMetadataCallback(printMetaData);
  player.begin();

  fileSize = source.getFileSize();
  progressStream.begin(fileSize);
}

void loop() {
  player.copy();

  Serial.print("Progress: ");
  Serial.print(progressStream.percentage());
  Serial.println("%");

  Serial.print("Filesize: ")
  Serial.println(fileSize);
  
  delay(1000);
}
```

However, the serial monitor still outputs the percentage as being zero, but the method I implemented is working because it successfully returns a value:

```
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
Progress: 0.00%
Filesize: 2801654
```

**Comment by pschatzmann** [comment]
That's because you still send the output to the I2SStream instead of the ProgressStream!
By the way you can also get the file size by calling static_cast<File*>(player.getStream())->size()...

Please note that your logic works only for the first file! you will need to repeat this when the file is changed.

**Reply by niuteam** [reply]
if not is SD File, then is UrlStream , the ProgressStream.setsize() is ?

**Reply by pschatzmann** [reply]
Depends, usually this is endless, so progress does not make any sense!
If the server is providing an http content-length, you would get it from the http reply header by calling contentLength()

**Reply by niuteam** [reply]
Thank you very much for your answer. I have already instructed the server to return the contentLength, which is the size of the MP3 file. I'm not sure whether the ProgressStream.setsize refers to the size of PCM or MP3. I tested, and neither of these values is correct. What does the value of progressStream.processedBytes indicate?

**Reply by pschatzmann** [reply]
Make sure that ProgressStream is before the decoding in the chain!

**Reply by niuteam** [reply]
I2SStream i2s;
MP3DecoderHelix decoder;
URLStream urlStream(urlStream, "audio/mp3");
ProgressStream progressStream(i2s);
AudioPlayer player(source, progressStream, decoder);


 player.begin();
 const char * len = urlStream.getReplyHeader("Content-Length");
 progressStream.setSize(atoi(len));


void Audio::loop()
{
    player.copy();  
    Serial.printf("[Audio] progress：%d\n", progressStream.percentage);
    vTaskDelay(pdMS_TO_TICKS(1));
}


Based on the code above, the actual situation is that the progress is completely incorrect?

**Reply by niuteam** [reply]
Looking forward to your reply. Thank you very, very much

**Reply by pschatzmann** [reply]
No,  the output for the AudioPlayer is decoded PCM data!
So build your own output chain instead of the AudioPlayer or use the setCallbackOnWrite() from the AudioPlayer's StreamCopy

**Comment by Sparky54** [comment]
> That's because you still send the output to the I2SStream instead of the ProgressStream!

I'm not quite sure what you mean. How would I correct this?

**Comment by pschatzmann** [comment]
In the constructor of the AudioPlayer you need to define the correct output:
ProgressStream progressStream(i2s);
AudioPlayer player(source, progressStream, decoder);

**Reply by Sparky54** [reply]
Awesome! Thank you!
